src/core/lifetime.d

```diff
@@ -2292,3 +2353,186 @@ void _d_delstruct(T)(ref T *p)
+
+    auto a = NotNested(Nested());
+    assert(a.n.tupleof[$-1]);
+    auto b = move(a);
+    assert(b.n.tupleof[$-1]);
+    assert(a.n.tupleof[$-1] is b.n.tupleof[$-1]);
+
+    auto c = Deep(NotNested(Nested()));
+    auto d = move(c);
+    assert(d.nn.n.tupleof[$-1]);
+    assert(c.nn.n.tupleof[$-1] is d.nn.n.tupleof[$-1]);
+
+    auto e = Deeper([NotNested(Nested())]);
+    auto f = move(e);
+    assert(f.nn[0].n.tupleof[$-1]);
+    assert(e.nn[0].n.tupleof[$-1] is f.nn[0].n.tupleof[$-1]);
+}
+
+// issue 25552
+@safe unittest
+{
+    int i;
+    struct Nested
+    {
+        int[3] stuff = [0, 1, 2];
+        void foo() { ++i; }
+        ~this() {}
+    }
+
+    static struct NoAssign
+    {
+        int value;
+        @disable void opAssign(typeof(this));
+    }
+
+    static struct NotNested
+    {
+        int before = 42;
+        Nested n;
+        auto after = NoAssign(43);
+    }
+
+    static struct Deep
+    {
+        NotNested nn;
+    }
+
+    static struct Deeper
+    {
+        NotNested[1] nn;
+    }
+
+    static assert(!__traits(isZeroInit, Nested));
+    static assert(!__traits(isZeroInit, NotNested));
+    static assert(!__traits(isZeroInit, Deep));
+    static assert(!__traits(isZeroInit, Deeper));
+
+    auto a = NotNested(1, Nested([3, 4, 5]), NoAssign(2));
+    auto b = move(a);
+    assert(b.n.tupleof[$-1]);
+    assert(a.n.tupleof[$-1] is b.n.tupleof[$-1]);
+    assert(a.n.stuff == a.n.init.stuff);
+    assert(a.before == a.init.before);
+    assert(a.after == a.init.after);
+
+    auto c = Deep(NotNested(1, Nested([3, 4, 5]), NoAssign(2)));
+    auto d = move(c);
+    assert(d.nn.n.tupleof[$-1]);
+    assert(c.nn.n.tupleof[$-1] is d.nn.n.tupleof[$-1]);
+    assert(c.nn.n.stuff == c.nn.n.init.stuff);
+    assert(c.nn.before == c.nn.init.before);
+    assert(c.nn.after == c.nn.init.after);
+
+    auto e = Deeper([NotNested(1, Nested([3, 4, 5]), NoAssign(2))]);
+    auto f = move(e);
+    assert(f.nn[0].n.tupleof[$-1]);
+    assert(e.nn[0].n.tupleof[$-1] is f.nn[0].n.tupleof[$-1]);
+    assert(e.nn[0].n.stuff == e.nn[0].n.init.stuff);
+    assert(e.nn[0].before == e.nn[0].init.before);
+    assert(e.nn[0].after == e.nn[0].init.after);
+}
+
+// wipes source after moving (zero init case)
+pragma(inline, true)
+private void wipe(T)(return scope ref T source) @trusted
+{
+    static if (__traits(isStaticArray, T) && hasContextPointers!T)
+    {
+        for (auto i = 0; i < T.length; i++)
+            wipe(source[i]);
+    }
+    else static if (is(T == struct) && hasContextPointers!T)
+    {
+        import core.internal.traits : anySatisfy;
+        static if (anySatisfy!(hasContextPointers, typeof(T.tupleof)))
+        {
+            static foreach (i; 0 .. T.tupleof.length - __traits(isNested, T))
+                wipe(source.tupleof[i]);
+        }
+        else
+        {
+            static if (__traits(isNested, T))
+                enum sz = T.tupleof[$-1].offsetof;
```

Comment:
Thx for the `.offsetof` fix - the previous code assumed the context pointer to be right at the end, without any tail padding.

I haven't checked, but I assume some nested `align(64) struct S { int x; }` to contain the pointer at offset 8 [for a 64-bit target], which would leave the 48 bytes tail padding undefined. I guess the safe variant would be to save the ptr, memset/cpy the whole initializer, and then restore the ptr again. Otherwise equality/`is` identity of structs could be broken.

Reply:
Ouch, and the recursive field-wise wiping is missing *all* paddings. :/

Reply:
Maybe this could/should be simplified to:
1. blit src into dst
2. reset src to T.init, clearing all context pointers but initializing all gaps and paddings
3. if need be: recurse into fields and copy context pointers back from dst to src

Reply:
Erm wait, this is the wiping of a supposedly valid instance - all paddings should be well-defined already. Please ignore. ;)

Reply:
> some nested `align(64) struct S { int x; }`

Oh yeah, need to add a test for things with custom-aligned pointer, thanks!